src/Data.Json.JTable.Internal.purs

```diff
@@ -0,0 +1,214 @@
+mergeObjTuple ::Tree -> JCursor -> [Json] -> Maybe Table
+mergeObjTuple t@(T p w h k) c ja = head ja *> do
+  jos <- for ja toObject
+  let keyss = jos <#> M.keys
+  let all_keys = concat keyss
+  so ((length all_keys) == (length $ nub all_keys)) $ 
+    cMergeObj $ k <#> \(t'@(T p' w' _ _)) -> let 
+      label = AU.last p'
+      i = findIndex (elem label) keyss
+      j = jnull `fromMaybe` (jos !! i >>= M.lookup label)
+      in Tuple w' (cFromJson t' (downField label $ downIndex i c) j)
+
+-- produce data table from json, according to header tree
+cFromJson :: Tree -> JCursor -> Json -> Table
+cFromJson t@(T p w h k) c json = let
+  obj = toObject json <#> \jo ->
+    if M.isEmpty jo then [[C c w 1 primNull]] 
+    else cMergeObj $ do
+      t'@(T p' w' _ _) <- k
+      let label = AU.last p'
+      let j = jnull `fromMaybe` M.lookup label jo
+      return $ Tuple w' (cFromJson t' (downField label c) j)
+  ma = toArray json
+  obtup = ma >>= mergeObjTuple t c
+  width = ma >>= widthOfPrimTuple p >>= so (h <= 0 && w > 1)
+  tuple = ma <#> \a -> singleton $ (0 .. w-1) <#> \i ->
+    C (downIndex i c) 1 1 $ toPrim $ jnull `fromMaybe` (a !! i)
+  array = ma <#> \a -> zipWithIndex a >>= uncurry
+                  \j i -> cFromJson t (downIndex i c) j
+  prim = [[C c w 1 $ toPrim json]]
+  in prim `fromMaybe` (obj <|> obtup <|> (width *> tuple) <|> array)
+
+
+-- render a grid from an array of arrays
+renderRows :: forall a. (Markup -> Markup) -> (Number -> Number -> a -> Markup) -> [[a]] -> Markup
+renderRows tr' cellf rows = mconcat $ do
+  (Tuple row y) <- zipWithIndex rows
+  return $ tr' $ mconcat $ do
+    (Tuple cell x) <- zipWithIndex row
+    return $ cellf y x cell
+
+_nattr :: String -> Number -> Markup -> Markup
+_nattr attr n m = if n > 1 then m ! (attribute attr $ show n) else m
+_cspan = _nattr "colspan"
+_rspan = _nattr "rowspan"
+
+-- produce header rows from header tree
+tsToRows :: [Tree] -> [[Tree]]
+tsToRows ts = if null ts then [] else ts : tsToRows (ts >>= tKids)
+
+renderThead :: (Markup -> Markup) -> (Tree -> Markup) -> Tree -> Markup
+renderThead tr' thf (T p w h k) =
+  let rs i k = if null k then h - i else 1
+      tdf' y x t@(T p w h k) = thf t # _cspan w >>> (_rspan $ rs y k)
+  in renderRows tr' tdf' $ tsToRows k
+
+renderTbody :: (Markup -> Markup) -> (Cell -> Markup) -> Tree -> Json -> Markup
+renderTbody tr' tdf t json =
+  let tdf' y x cell@(C c w h j) = tdf cell # _cspan w >>> _rspan h
+  in renderRows tr' tdf' $ cFromJson t JCursorTop json
+
+-- sort header tree by ColumnOrdering
+sortTree :: (JPath -> JPath -> Ordering) -> Tree -> Tree
+sortTree ord (T p w h k) = T p w h ( 
+  sortBy (\t1 t2 -> ord (t1 # tPath) (t2 # tPath)) (k <#> sortTree ord))
+
+
+strcmp :: String -> String -> Ordering
+strcmp s1 s2 = compare (localeCompare s1 s2) 0
+
+
+-- pad tall header cells from above
+insertHeaderCells :: Number -> Tree -> Tree
+insertHeaderCells maxh t@(T p w h k) = 
+  if null k 
```

Comment:
The pattern matching version would look like this:
```
insertHeaderCells maxh t@(T p w h k) = case k of
  (_:_) -> T p w h (k <#> insertHeaderCells (maxh - 1))
  _     -> if maxh > 0 then T [] w 1 [insertHeaderCells (maxh - 1) t]
                       else T p w 1 k
```
Of which the pattern match translates to
```
if (_33.value3.length >= 1) {
  var _120 = _33.value3.slice(1); 
/* ... */}
```
This would not only have the same length check but also copy the array, only to discard the copy. Usin pattern matching here doesn't bring any benefit, because no inferred information is used, it's just trying to keep up List-like appearances, but in the end we're working with Arrays, not Lists.